### PR TITLE
Initial work to constrain when the `tk_ai` (anon ID tracking cookie) is set

### LINKED ISSUE #40688: Tracking cookie (`tk_ai`) is being set even when tracking is disabled.
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

According to the [WooCommerce Cookies](https://woocommerce.com/document/woocommerce-cookies/) guide, the `tk_ai` cookie should only be set when usage tracking in enabled, however, that's not the case on all of my test sites. 

### Expected behavior

The cookie to be stored only when the merchant opt-in for tracking. 

### Actual behavior

The cookie being set regardless.

### Steps to reproduce

1. Navigate to any admin page while disabling usage tracking from `WooCommerce > Settings > Advance > WooCommerce.com`.

Link to image: https://d.pr/i/sgWJlT

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): http://woo-cookie-test.local
Site address (URL): http://woo-cookie-test.local
WC Version: 8.2.0
REST API Version: ✔ 8.2.0
WC Blocks Version: ✔ 11.1.2
Action Scheduler Version: ✔ 3.6.3
Log Directory Writable: ✔
WP Version: 6.3.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 8.1.23
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 8.1.2
(SecureTransport) LibreSSL/3.3.6

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.2.0
WC Database Prefix: wp_
Total Database Size: 5.26MB
Database Data Size: 3.60MB
Database Index Size: 1.66MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.52MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.05MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
page: 7
post: 2
wp_navigation: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 8.2.0

### Settings ###

API Enabled: –
Force SSL: –
Currency: JOD (د.ا)
Currency Position: right_space
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 3
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/
Checkout: #9 - /checkout/
My account: #10 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Three
Version: 1.2
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2023-10-11 22:38:19 +00:00
Options: ✔
Notes: 61
Onboarding: skipped

### Action Scheduler ###

Complete: 2
Oldest: 2023-10-10 22:39:54 +0000
Newest: 2023-10-10 22:39:54 +0000

Pending: 1
Oldest: 2023-10-11 22:39:54 +0000
Newest: 2023-10-11 22:39:54 +0000

### Status report information ###

Generated at: 2023-10-10 22:53:30 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by barryhughes**
Addresses concerns that the `tk_ai` (anonymous identifier tracking cookie) is being set, even when tracking has been disabled.

Closes #40688.

### How to test the changes in this Pull Request:

<div align="center">Setting to enable or disable tracking</div>

1. Visit **WooCommerce ‣ Settings ‣ Advanced ‣ WooCommerce.com** and enable tracking (remember to save your changes each time you modify this!).
2. Use your browser's dev tools to ensure that a cookie named `tk_ai` has been set.
3. Delete the cookie and reload the page, or navigate to any other settings page. The cookie should have been restored.
4. Now disable tracking, and repeat the above test (delete the cookie, and reload the page or navigate to another admin page). This time, the cookie should *not* be restored.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by barryhughes** [comment]
This may not be required: it's not clear that the problem surfaces using WooCommerce alone.

**Comment by barryhughes** [comment]
~Closing, as this may not be needed.~ Edit: still required.

File: plugins/woocommerce/includes/tracks/class-wc-site-tracking.php [hunk 1]
```diff
@@ -233,5 +233,26 @@ public static function init() {
...
+	 * @param string $cookie_value The value of the cookie.
+	 * @param int    $expire       Expiry of the cookie.
+	 * @param bool   $secure       Whether the cookie should be served only over https.
+	 * @param bool   $http_only    Whether the cookie is only accessible over HTTP.
+	 *
+	 * @return bool If setting the cookie was attempted (will be false if tracking is not allowed).
+	 */
+	public static function set_tracking_cookie( string $cookie_key, string $cookie_value, int $expire = 0, bool $secure = false, bool $http_only = false ): bool {
+		if ( self::is_tracking_enabled() ) {
+			wc_setcookie( $cookie_key, $cookie_value, $expire, $secure, $http_only);
```
Comment (barryhughes) [inline plugins/woocommerce/includes/tracks/class-wc-site-tracking.php]:
```suggestion
			wc_setcookie( $cookie_key, $cookie_value, $expire, $secure, $http_only );
```

Coding standards.

**Review by naman03malhotra — COMMENTED** [review]
Thanks for working on it! Looks good.

The code change works as expected.

However, a good followup can be to remove the cookie when the tracking option is unchecked. Right now it is manual, but that is intended for this part of the fix.

image


image

File: plugins/woocommerce/includes/tracks/class-wc-site-tracking.php [hunk 2]
```diff
@@ -233,5 +233,26 @@ public static function init() {
...
+	 *
+	 * @param string $cookie_key   The key of the cookie.
+	 * @param string $cookie_value The value of the cookie.
+	 * @param int    $expire       Expiry of the cookie.
+	 * @param bool   $secure       Whether the cookie should be served only over https.
+	 * @param bool   $http_only    Whether the cookie is only accessible over HTTP.
+	 *
+	 * @return bool If setting the cookie was attempted (will be false if tracking is not allowed).
+	 */
+	public static function set_tracking_cookie( string $cookie_key, string $cookie_value, int $expire = 0, bool $secure = false, bool $http_only = false ): bool {
```
Comment (naman03malhotra) [inline plugins/woocommerce/includes/tracks/class-wc-site-tracking.php:250]: Please consider writing a unit test for this change.

**Comment by barryhughes** [comment]
> A good followup can be to remove the cookie when the tracking option is unchecked.

This is probably safe to do if we can be sure it was set by Woo, and there does seem to be a prefix that would allow us to determine that, at least in most cases (...the complicating factor being that Jetpack could also be present, and it also sets a cookie with the same key for the same sort of purpose).

**Comment by naman03malhotra** [comment]
>at least in most cases (...the complicating factor being that Jetpack could also be present, and it also sets a cookie with the same key for the same sort of purpose).

Hmm, in that case, we can open a follow up issue and see if can find a better way to deal with this, but for now this should work.

File: plugins/woocommerce/includes/tracks/class-wc-site-tracking.php (hunk 2 above)
Comment (barryhughes) [reply plugins/woocommerce/includes/tracks/class-wc-site-tracking.php:250]: https://github.com/woocommerce/woocommerce/pull/47863/commits/fd02be5f4be483ecdf109dd707f4a36ed6530ec7

**Review by naman03malhotra — APPROVED** [review]
Looks good, thanks for adding test. 🏁

**Pull request merged**